builder.py

```diff
@@ -17,45 +17,115 @@
-    if config_settings is None:
-        config_settings = {}
+def pkg_config_lookup_var(lib, varname) -> str:
+    full_cmd = ["pkg-config", lib, f"--variable={varname}"]
+    # print the command for transparency
+    print(" ".join(full_cmd))
+    return subprocess.check_output(full_cmd).decode().strip()
+
+
+def pkg_config_flags(*libraries: str):
+    include_dirs = []
+    library_dirs = []
+
+    for libname in libraries:
+        library_dirs.append(pkg_config_lookup_var(libname, "libdir"))
+
+        inc_dir = pkg_config_lookup_var(libname, "includedir")
+
+        if (subdir := Path(inc_dir) / libname).exists():
+            inc_dir = subdir
+
+        include_dirs.append(inc_dir)
+
+    libs = [lib.removeprefix("lib") for lib in libraries]
+    return include_dirs, library_dirs, libs
+
+
+def prepare_extensions():
+    includes, libdirs, libs = pkg_config_flags(
+        "wlroots-0.19",
+        "wayland-server",
+        "cairo",
+        "pixman-1",
+        "libdrm"
+    )
+
+    wayland_backend = Extension(
+        "wayland_backend",
+        sources=[str(filepath) for filepath in WAYLAND_BACKEND_DIR.rglob("*.c")],
+        language="c",
+        library_dirs=libdirs,
+        libraries=list(libs),
+        include_dirs=[str(WAYLAND_BACKEND_DIR), str(QW_PROTO_OUT_PATH)] + includes,
+        extra_compile_args=["-DWLR_USE_UNSTABLE"]
+    )
+
+    return [wayland_backend]
+
+
+def build_cff_lib(config_settings: dict[str, str]) -> None:
+    wayland_requested = any(
+        config_settings.get(key, "").lower() == "wayland"
+        for key in ("Backend", "backend")
+    )
 
-    wayland_requested = wants_wayland(config_settings)
     try:
         from libqtile.backend.wayland.cffi.build import ffi
 
+        print("Building qw cffi")
         ffi.compile(verbose=wayland_requested)
+
     except Exception as e:
         if wayland_requested:
             sys.exit(f"Wayland backend requested but backend could not be built: {e}")
         else:
             print("Wayland backend was not built.")
 
-    # Write library paths to file, if they are specified at build time via
-    # --config-settings=PANGO_PATH=...
-    with open("libqtile/_build_config.py", "w") as f:
-        f.write("# This file is generated at build time by builder.py\n")
-        for lib in ["PANGO", "PANGOCAIRO", "GOBJECT", "XCBCURSOR"]:
-            p = lib + "_PATH"
-            f.write(f"{p} = {config_settings.get(p)!r}\n")
+
+def build_wheel(wheel_directory, config_settings=None, metadata_directory=None):
+    if config_settings is None:
+        config_settings = {}
+
+    from setuptools.command.build_ext import build_ext
+
+    generate_build_config(config_settings or {})
+    build_cff_lib(config_settings or {})
+
+    dist = Distribution()
+    dist.ext_modules = prepare_extensions()
+
+    print("Building Wayland backend extension")
+
+    cmd = build_ext(dist)
+    cmd.build_lib = WAYLAND_DIR
+    cmd.ensure_finalized()
+    cmd.run()
```

Comment:
What happens if the build fails? `build_wheel` will be run if people just want the x11 backend so we must allow this function to succeed.

My previous version only raised an error if the wayland backend didn't build when the user explicitly requests the backend via `--config-settings backend=wayland`. I think you need that here unless I'm missing something.